### DISCUSSION TITLE
More info on tracking one-time charges

### DISCUSSION THREAD

**Discussion by harrison-broadbent** [pay-rails/pay#938] [Q&A]
Hey! 

I'm curious what the recommended way to handle one-time purchases through `pay` is. Specifically, I'm looking at Section 8 of the Stripe docs (https://github.com/pay-rails/pay/blob/main/docs/stripe/8_stripe_checkout.md#fulfilling-orders-after-checkout-completed) where there's some information on fulfilling orders after a checkout. 

I'm struggling to figure out what should go in the `# Handle fulfillment` section of the given code example — 

```
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.async_payment_succeeded", FulfillCheckout.new

class FulfillCheckout
  def call(event)
    object = event.data.object

    return if object.payment_status != "paid"

    # Handle fulfillment
  end
end
```

I'm wondering if I should manually create a `Pay::Subscription` here to indicate that the User (who just purchased access or whatever) now has access to the product. 

Does this seem right, or am I missing something? I'm wondering if there's a state on `Pay::Customer` I should set here instead; I've gone through the `Pay::Customer` code at https://github.com/pay-rails/pay/blob/main/app/models/pay/customer.rb but nothing stood out to me. 

Thanks!

**Comment by excid3** [comment]
Yeah, although webhooks aren't reliable so you could end up in a situation where the user is redirected before the fulfillment happens. 

I also use the redirect_url from the checkout session to call `Pay::Stripe::Subscription.sync_from_checkout_session` to ensure that it's fulfilled even if webhooks are delayed. This is helpful as well that you don't _need_ to have Stripe CLI running in development for basic checkout testing. The user isn't guaranteed to hit this URL I've noticed, so the combination of this and webhooks doing fulfillment should handle every situation.

**Reply by excid3** [reply]
We were just talking yesterday about introducing a `Pay::Stripe.sync_from_checkout_session` that can handle subscriptions or one-time payments to simplify the interface for users. Gonna hack on that today if I get a chance.

**Reply by harrison-broadbent** [reply]
@excid3 Thanks for the info! 🙌 Appreciate it and would love to hear how you go today hacking on that stuff 👍

**Reply by excid3** [reply]
Started this: https://github.com/pay-rails/pay/pull/943

**Reply by harrison-broadbent** [reply]
@excid3 🙏
